📖 [02] TIỀN XỬ LÝ DỮ LIỆU - PIMA INDIANS DIABETES

Mục tiêu: Làm sạch và chuẩn bị dữ liệu cho bước huấn luyện mô hình bằng cách xử lý giá trị thiếu, thay thế giá trị không hợp lệ và chuẩn hóa dữ liệu.

Các bước chính:

1. Đọc và kiểm tra dữ liệu

- Tải file diabetes_structured.csv.

- Kiểm tra kích thước, giá trị thiếu và số lượng giá trị 0 trong các cột quan trọng.

2. Làm sạch dữ liệu

- Thay thế các giá trị 0 không hợp lệ (ở các cột Glucose, BloodPressure, SkinThickness, Insulin, BMI) bằng giá trị NaN.

- Xác nhận lại số lượng giá trị thiếu sau khi thay thế.

3. Tạo các bộ dữ liệu xử lý khác nhau

- Dataset 1 – Mean: Thay thế các giá trị thiếu bằng mean của từng cột.

- Dataset 2 – Median: Thay thế các giá trị thiếu bằng median của từng cột.

- Dataset 3 – Scaled: Chuẩn hóa dữ liệu từ Dataset 2 bằng StandardScaler để đưa các giá trị về cùng thang đo.

4. Lưu dữ liệu đã xử lý

- Lưu ba file kết quả:

  - diabetes_dataset1_mean.csv

  - diabetes_dataset2_median.csv

  - diabetes_dataset3_scaled.csv

5. Kết luận

- Dữ liệu sau khi tiền xử lý không còn giá trị thiếu hoặc không hợp lệ.

- Bộ dữ liệu đã được chuẩn bị ở ba mức độ khác nhau để thử nghiệm với các mô hình học máy: Mean, Median và Scaled.

1.1. Import thư viện và đọc dữ liệu

In [1]:
# ===========================
# ĐỌC DỮ LIỆU
# ===========================
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler

# Đọc file dữ liệu
df = pd.read_csv('diabetes_structured.csv')

print(f"Kích thước dữ liệu: {df.shape[0]} dòng, {df.shape[1]} cột")
print(df.head())
print(df.info())

Kích thước dữ liệu: 768 dòng, 9 cột
   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64 

1.2. Kiểm tra dữ liệu ban đầu

In [2]:
# ===========================
#  KIỂM TRA DỮ LIỆU
# ===========================
print("\nKiểm tra giá trị thiếu:")
print(df.isnull().sum())

print("\nKiểm tra số lượng giá trị 0 trong các cột:")
zero_cols = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']
for col in zero_cols:
    if col in df.columns:
        zero_count = (df[col] == 0).sum()
        print(f"  • {col:<20}: {zero_count:>3} giá trị 0 ({zero_count/len(df)*100:>5.2f}%)")



Kiểm tra giá trị thiếu:
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

Kiểm tra số lượng giá trị 0 trong các cột:
  • Glucose             :   5 giá trị 0 ( 0.65%)
  • BloodPressure       :  35 giá trị 0 ( 4.56%)
  • SkinThickness       : 227 giá trị 0 (29.56%)
  • Insulin             : 374 giá trị 0 (48.70%)
  • BMI                 :  11 giá trị 0 ( 1.43%)


**Nhận xét:**

Các cột không có giá trị NaN, nhưng có nhiều giá trị bằng 0 ở các cột không hợp lý:

- Glucose: 5 giá trị 0

- BloodPressure: 35 giá trị 0

- SkinThickness: 227 giá trị 0

- Insulin: 374 giá trị 0

- BMI: 11 giá trị 0

Đây là dấu hiệu dữ liệu thiếu cần xử lý.

2. Làm sạch dữ liệu (thay 0 bằng NaN)

In [3]:
# ===========================
# LÀM SẠCH DỮ LIỆU
# ===========================
df_clean = df.copy()

# Thay thế giá trị 0 bằng NaN trong các cột không thể có giá trị 0
for col in zero_cols:
    if col in df_clean.columns:
        df_clean[col] = df_clean[col].replace(0, np.nan)

print("\nĐã thay thế giá trị 0 bằng NaN trong các cột:", zero_cols)
print("\nSố lượng giá trị thiếu sau khi xử lý:")
print(df_clean.isnull().sum())



Đã thay thế giá trị 0 bằng NaN trong các cột: ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']

Số lượng giá trị thiếu sau khi xử lý:
Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64


**Nhận xét:**

Sau bước này, các giá trị 0 được thay bằng NaN:

- Glucose có 5 giá trị thiếu

- BloodPressure có 35 giá trị thiếu

- SkinThickness có 227 giá trị thiếu

- Insulin có 374 giá trị thiếu

- BMI có 11 giá trị thiếu

3.a. Xử lý giá trị thiếu bằng MEAN

In [4]:
# ===========================
#  DATASET 1 - THAY GIÁ TRỊ THIẾU BẰNG MEAN
# ===========================
print("\n" + "-"*60)
print("DATASET 1: Thay giá trị thiếu bằng MEAN")
print("-"*60)

df_mean = df_clean.copy()
for col in zero_cols:
    if col in df_mean.columns:
        mean_value = df_mean[col].mean()
        df_mean[col] = df_mean[col].fillna(mean_value)
        print(f"  - {col}: Thay NaN bằng mean = {mean_value:.2f}")

print(f"Số lượng giá trị thiếu còn lại: {df_mean.isnull().sum().sum()}")




------------------------------------------------------------
DATASET 1: Thay giá trị thiếu bằng MEAN
------------------------------------------------------------
  - Glucose: Thay NaN bằng mean = 121.69
  - BloodPressure: Thay NaN bằng mean = 72.41
  - SkinThickness: Thay NaN bằng mean = 29.15
  - Insulin: Thay NaN bằng mean = 155.55
  - BMI: Thay NaN bằng mean = 32.46
Số lượng giá trị thiếu còn lại: 0


Tất cả giá trị thiếu đã được thay thế bằng giá trị
trung bình

Không còn giá trị thiếu.

3.b. Xử lý giá trị thiếu bằng MEDIAN

In [6]:
# ===========================
# DATASET 2 - THAY GIÁ TRỊ THIẾU BẰNG MEDIAN
# ===========================
print("\n" + "-"*60)
print("DATASET 2: Thay giá trị thiếu bằng MEDIAN")
print("-"*60)

df_median = df_clean.copy()
for col in zero_cols:
    if col in df_median.columns:
        median_value = df_median[col].median()
        df_median[col] = df_median[col].fillna(median_value)
        print(f"  - {col}: Thay NaN bằng median = {median_value:.2f}")

print(f"Số lượng giá trị thiếu còn lại: {df_median.isnull().sum().sum()}")



------------------------------------------------------------
DATASET 2: Thay giá trị thiếu bằng MEDIAN
------------------------------------------------------------
  - Glucose: Thay NaN bằng median = 117.00
  - BloodPressure: Thay NaN bằng median = 72.00
  - SkinThickness: Thay NaN bằng median = 29.00
  - Insulin: Thay NaN bằng median = 125.00
  - BMI: Thay NaN bằng median = 32.30
Số lượng giá trị thiếu còn lại: 0


Toàn bộ NaN được thay bằng trung vị.

Không còn giá trị thiếu. Phương pháp này giúp dữ liệu ổn định hơn khi có outlier.

3.c. Chuẩn hóa dữ liệu (StandardScaler)

In [7]:
# ===========================
# DATASET 3 - CHUẨN HÓA DỮ LIỆU (TỪ DATASET 2)
# ===========================
print("\n" + "-"*60)
print("DATASET 3: Chuẩn hóa dữ liệu (từ Dataset 2)")
print("-"*60)

df_scaled = df_median.copy()

if 'Outcome' in df_scaled.columns:
    X_to_scale = df_scaled.drop('Outcome', axis=1)
    y = df_scaled['Outcome']
else:
    X_to_scale = df_scaled

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_to_scale)
X_scaled_df = pd.DataFrame(X_scaled, columns=X_to_scale.columns)

if 'Outcome' in df_scaled.columns:
    df_scaled_final = pd.concat([X_scaled_df, y.reset_index(drop=True)], axis=1)
else:
    df_scaled_final = X_scaled_df

print("Đã chuẩn hóa dữ liệu bằng StandardScaler!")
print(df_scaled_final.head())



------------------------------------------------------------
DATASET 3: Chuẩn hóa dữ liệu (từ Dataset 2)
------------------------------------------------------------
Đã chuẩn hóa dữ liệu bằng StandardScaler!
   Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0     0.639947  0.866045      -0.031990       0.670643 -0.181541  0.166619   
1    -0.844885 -1.205066      -0.528319      -0.012301 -0.181541 -0.852200   
2     1.233880  2.016662      -0.693761      -0.012301 -0.181541 -1.332500   
3    -0.844885 -1.073567      -0.528319      -0.695245 -0.540642 -0.633881   
4    -1.141852  0.504422      -2.679076       0.670643  0.316566  1.549303   

   DiabetesPedigreeFunction       Age  Outcome  
0                  0.468492  1.425995        1  
1                 -0.365061 -0.190672        0  
2                  0.604397 -0.105584        1  
3                 -0.920763 -1.041549        0  
4                  5.484909 -0.020496        1  


**Nhận xét:**

Các giá trị đã được chuẩn hóa về trung bình ≈ 0, độ lệch chuẩn ≈ 1.

Dữ liệu này sẵn sàng để huấn luyện các mô hình như KNN

4. Lưu dữ liệu đã xử lý

In [8]:
# ===========================
# LƯU DỮ LIỆU ĐÃ XỬ LÝ
# ===========================
print("\n" + "="*60)
print("LƯU DỮ LIỆU ĐÃ XỬ LÝ")
print("="*60)

df_mean.to_csv('diabetes_dataset1_mean.csv', index=False)
print("Đã lưu: diabetes_dataset1_mean.csv")

df_median.to_csv('diabetes_dataset2_median.csv', index=False)
print("Đã lưu: diabetes_dataset2_median.csv")

df_scaled_final.to_csv('diabetes_dataset3_scaled.csv', index=False)
print("Đã lưu: diabetes_dataset3_scaled.csv")

print("\nHOÀN THÀNH TIỀN XỬ LÝ DỮ LIỆU!")



LƯU DỮ LIỆU ĐÃ XỬ LÝ
Đã lưu: diabetes_dataset1_mean.csv
Đã lưu: diabetes_dataset2_median.csv
Đã lưu: diabetes_dataset3_scaled.csv

HOÀN THÀNH TIỀN XỬ LÝ DỮ LIỆU!


**Nhận xét:**

Ba tệp dữ liệu mới được tạo ra trong thư mục hiện tại:

- diabetes_dataset1_mean.csv

- diabetes_dataset2_median.csv

- diabetes_dataset3_scaled.csv

Mỗi tệp tương ứng với một phương pháp tiền xử lý khác nhau, có thể sử dụng để huấn luyện và so sánh độ chính xác giữa các mô hình.

5. Kết luận

  Quá trình tiền xử lý dữ liệu bao gồm ba bước chính: làm sạch giá trị sai lệch, xử lý giá trị thiếu bằng hai phương pháp (trung bình và trung vị), và chuẩn hóa dữ liệu bằng StandardScaler.

  Các bước này giúp loại bỏ nhiễu, giảm sai số, đồng thời đảm bảo rằng dữ liệu đầu vào có chất lượng tốt, phù hợp cho các mô hình học máy.

  Kết quả thu được là ba phiên bản dữ liệu khác nhau, giúp so sánh và lựa chọn cách tiền xử lý tối ưu nhất cho bài toán dự đoán tiểu đường.